content/posts/draft-operator-constraints.markdown

```diff
@@ -0,0 +1,349 @@
+
+I believe the fourth is a novel idea. At least I don't think I have seen it
+come up in a discussion about Go so far. Though I think I have seen it used in
+other languages.
+
+## `constraints.Ordered`
+
+Go 1.18 has a mechanism to constraint a type parameter to all types which have
+the `<` operator defined on them. The types which have this operator are
+exactly all types whose underlying type is `string` or one of the predeclared
+integer and float types. So we can write a type set expressing that:
+
+```go
+type Integer interface {
+  ~int | ~int8 | ~int16 | ~int32 | ~int64 | ~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr
+}
+
+type Float interface {
+  ~float32 | ~float64
+}
+
+type Ordered interface {
+  Integer | Float | ~string
+}
+```
+
+Because that's a fairly common thing to want to do,
+[there is already a package which contains these kinds of type sets](https://pkg.go.dev/golang.org/x/exp/constraints#Ordered).
+
+With this, you can write the signature of your sorting function or the
+definition of your search tree as:
+
+```go
+func Sort[T constraints.Ordered](s []T) {
+  // …
+}
+
+type SearchTree[T constraints.Ordered] struct {
+  // …
+}
+```
+
+The main advantage of this is that it works directly with predeclared types and
+simple types like `time.Duration`. It also is very clear.
+
+The main disadvantage is that it does not allow composite types like `struct`s.
+And what if a user want a different sorting order than the one implied by `<`?
+For example if they want to reverse the order or want specialized string
+collation. A multimedia library might want to sort “The Expanse” under E. And
+some letters sort differently depending on the language setting.
+
+`constraints.Ordered` is simple, but it also is inflexible.
+
+## Method constraints
+
+We can use method constraints to allow more flexibility. This allows a user to
+implement whatever sorting order they want as a method on their type.
+
+We can write that constraint like this:
+
+```go
+type Lesser[T any] interface {
+  // Less returns if the receiver is less than v.
+  Less(v T) bool
+}
+```
+
+The type parameter is necessary because we have to refer to the receiver type
+itself in the `Less` method. This is hopefully clearer when we look at how this
+is used:
+
+```go
+func Sort[T Lesser[T]](s []T) {
+  // …
+}
+
+func SearchTree[T Lesser[T]](s []T) {
+  // …
+}
+```
+
+This allows the user of our library to customize the sorting order by defining
+a new type with a `Less` method:
+
+```go
+type ReverseInt int
+
+func (i ReverseInt) Less(j ReverseInt) bool {
+  return j < i // order is reversed
+}
+```
+
+The disadvantage of this is that it requires some boiler plate on part of your
+user. Using a custom sorting order always requires defining a type with a method.
+
+They can't use your code with predeclared types like `int` or `string` but
+always have to wrap it into a new type.
+
+Likewise, if a type already has a natural comparison method but it is not
```

Comment:
This comma feels out of place when the sentence is so short. Either remove the comma or merge the first and second sentence.
```suggestion
Likewise if a type already has a natural comparison method but it is not
```